api/envoy/api/v2/route/route.proto

```diff
@@ -491,7 +501,18 @@ message CorsPolicy {
   core.RuntimeFractionalPercent shadow_enabled = 10;
 }
 
-// [#comment:next free field: 30]
+// [#not-implemented-hide:]
+message ClusterSource {
+  // The cluster name.
+  string name = 1 [(validate.rules).string.min_bytes = 1];
+
+  // If specified, indicates a server to contact to get the cluster data via CDS.
+  // By default, Envoy will know where to get CDS data from via its bootstrap file, but
+  // other clients (e.g., gRPC) may not have a default.
```

Comment:
I guess the question I have here is why does gRPC not have a default CDS? I can see how this feature might be possibly useful in the general case, but it seems like gRPC should just be able to deal with this via its own bootstrap.

Reply:
I think this is one of those cases where a proxy like Envoy and a proxyless client like gRPC will have a slightly different world-view.

As a proxy, Envoy wants to get data for all clusters, since it does not know in advance which of them it will need to use.  The set of data it's interested in from CDS is not altered by what it may get via LDS or RDS.  Therefore, Envoy is expected to make a "wildcard" query for CDS (i.e., it does not populate the `DiscoveryRequest.resource_names` field), so that it gets back data for all clusters.  As a result, it makes sense to configure CDS in Envoy's bootstrap file, the same way that LDS is handled.  Effectively, there are two different configuration "roots" that Envoy loads automatically at startup: LDS and CDS.

In contrast, a proxyless client like gRPC is only ever interested in a single virtual host, which means that it does not care about *all* clusters; it wants to see only those clusters that are used by the particular virtual host that it's trying to contact.  As a result, instead of making a wildcard CDS query at startup, gRPC will wait for the RDS response and then make a CDS query for just the clusters it found in the RDS response.  In effect, there is only one config "root" for gRPC, which is LDS; everything else is "chained" out from that point (LDS -> RDS -> CDS -> EDS).

Given that, it seems inconsistent to have CDS configured via gRPC's bootstrap file.  The linkage from LDS to RDS and from CDS to EDS already happens via a ConfigSource.  My intent in this PR is to make this pattern consistent so that the whole chain works the same way.

In addition, it seems like this is something we'd want to do anyway to support federation.  How else would we delegate a particular set of clusters to a separate set of management servers?  It seems like this is consistent with the approach we're trying to take for UDPA, so adding this capability here seems like it moves us closer to the direction we eventually want to go in anyway.

Reply:
BTW, to be clear, I am not expecting this to break any of Envoy's existing functionality.  Envoy can continue to specify the CDS server in its bootstrap file.  These new ConfigSource fields in RDS would be optional, and if not set, the client can fall back to some reasonable default (Envoy could use the CDS server in its bootstrap file; gRPC would probably default to the same server we're already talking to for RDS).

As a side note, I think this is another case where the self-referential mechanism in #8201 would be useful.

Reply:
Yeah, this perspective is reasonable for gRPC, but the main concern that remains for me is that if we do add this as new behavior to Envoy, this is actually a lot of implementation complexity in `ClusterManager`; it needs to deal with multiple CDS configuration sources and to dynamically update these based on RDS. It's hard to see Envoy taking this as an implementation change within v2/v3 API lifetimes without some significant use case to justify this.